src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1436,15 +1540,29 @@ public bool CleanCacheAndDisposeIfUnused()
                 // Get the current time.  This is compared against each connection's last returned
                 // time to determine whether a connection is too old and should be closed.
                 long nowTicks = Environment.TickCount64;
-                Http2Connection? http2Connection = _http2Connection;
+                // Copy the reference to a local variable to simplify the removal logic below.
+                Http2Connection[]? localHttp2Connections = _http2Connections;
 
-                if (http2Connection != null)
+                if (localHttp2Connections != null)
                 {
-                    if (http2Connection.IsExpired(nowTicks, pooledConnectionLifetime, pooledConnectionIdleTimeout))
+                    for (int i = 0; i < localHttp2Connections.Length; i++)
```

Comment:
There's a bunch of similar logic below for dealing with HTTP1.1 connection in pretty much the same manner that we're doing here. The HTTP1.1 code has some advantages, such as ensuring the new array is only allocated/copied once. 

I don't know that we need to generalize the HTTP1.1 code, or exactly mimic it, etc, but it seems worth looking at what's done there and seeing if we should do something similar here.

Reply:
New efficient version of removal algorithm is implemented. It allocates only one array and only when it's necessary. It seems too hard (if even possible) to generalize HTTP 1.1 code here because it modifies the existing HTTP 1.1 connection list where `_http2Connections` is treated as immutable.